# Cosine Similarity between 2 sentence

In [1]:
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

text1 = "This is a foo bar sentence ."
text2 = "This sentence is similar to a foo bar sentence ."

tfidf = TfidfVectorizer(stop_words='english')
data = tfidf.fit_transform([text1,text2])

vector1 = tfidf.transform([text1]).toarray()
vector2 = tfidf.transform([text2]).toarray()

cosine = cosine_similarity(vector1, vector2)
print("Cosine:", cosine)

Cosine: [[0.81775832]]


# Cosine Similarity between 2 Document

In [5]:
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

doc1 = """
Coronavirus India State Wise Cases Latest Updates, Lockdown 5 News:  India has reported the highest single-day spike of 9,987 new COVID-19 cases and 331 deaths in the last 24 hours. The total number of coronavirus cases in the country now stands at 2,66,598. There are 1,29,917 active cases and 1,29,215 people have been cured/discharged/migrated, the Ministry of Health and Family Welfare latest data released on Tuesday said. The death toll has reached 7,466.  Meanwhile, Delhi Deputy Chief Minister Manish Sisodia has said that the city will 5.5 lakh coronavirus cases by July 31 of the current doubling rate continues. “5.5 lakh COVID-19 cases are expected in Delhi by July 31. We will need 80,000 beds by then,” Sisodia said after a meeting with Lieutenant Governor Anil Baijal. Sisodia also said that as per the Central government officials there is no community spread of coronavirus in Delhi so far.
Coronavirus cases in India have breached the 2.66-lakh-mark and with over 1.29 lakh people being cured of the highly contagious disease, India’s recovery rate is said to be one of the best in the world. However, the World Health Organisation (WHO) has warned that the COVID-19 pandemic is ‘worsening’ globally. It said that while the situation in Europe, which witnessed thousands of deaths due to coronavirus in past few months, is improving, American, South Asian and African nations are witnessing high number of cases daily. WHO chief Tedros Adhanom Ghebreyesus said there are encouraging signs from several countries , but added that ‘complacency’ could be a big threat.
"""

doc2 = """
A senior officer of the CBI in the Kolkata unit tested positive for COVID-19, officials said Tuesday.

The officer, who is in a supervisory role looking after anti-corruption cases of the CBI’s Kolkata unit has become first such case among senior officers of the probe agency, they said.
Sources in the agency said all possible medical help has been extended to the family.

India has become fifth worst-hit nation by the COVID-19 pandemic after the US, Brazil, Russia and the UK, according to the Johns Hopkins University data.

The country has registered over 9,000 coronavirus infection cases for the sixth day in a row taking the country tally to 2,66,598, according to Health Ministry data on Tuesday morning.
Get live Stock Prices from BSE, NSE, US Market and latest NAV, portfolio of Mutual Funds, calculate your tax by Income Tax Calculator, know market’s Top Gainers, Top Losers & Best Equity Funds. Like us on Facebook and follow us on Twitter.
"""

doc3 = """
Dozens of scientists doing research funded by Mark Zuckerberg say Facebook should not be letting President Donald Trump use the social media platform to “spread both misinformation and incendiary statements.”

The researchers, including 60 professors at leading US research institutions, wrote the Facebook CEO on Saturday asking Zuckerberg to “consider stricter policies on misinformation and incendiary language that harms people,” especially during the current turmoil over racial injustice.


 
The letter calls the spread of “deliberate misinformation and divisive language” contrary to the researchers’ goals of using technology to prevent and eradicate disease, improve childhood education and reform the criminal justice system.


Their mission “is antithetical to some of the stances that Facebook has been taking, so we’re encouraging them to be more on the side of truth and on the right side of history as we’ve said in the letter,” said Debora Marks of Harvard Medical School, one of three professors who organized it.

The others are Martin Kampmann of the University of California-San Francisco and Jason Shepherd of the University of Utah. All have grants from a Chan Zuckerberg Initiative program working to prevent, cure and treat neurodegenerative disorders including Alzheimer’s and Parkinson’s disease.

They said the letter had more than 160 signatories. Shepherd said about 10% are employees of foundations run by Zuckerberg and his wife, Priscilla Chan.


 
The letter objects specifically to Zuckerberg’s decision not to at least flag as a violation of Facebook’s community standards Trump’s post that stated “when the looting starts, the shooting starts” in response to unrest in Minneapolis over the videotaped killing of George Floyd, a black man, by a white police officer. The letter’s authors called the post “a clear statement of inciting violence.”

Twitter had both flagged and demoted a Trump tweet using the same language.

In a statement, the Chan Zuckerberg Initiative noted that the philanthropic organization is separate from Facebook and said “we are grateful for our staff, partners and grantees” and “respect their right to voice their opinions, including on Facebook policies.”

Some Facebook employees have publicly objected to Zuckerberg’s refusal to take down or label misleading or incendiary posts by Trump and other politicians. But Zuckerberg — who controls a majority of voting shares in the company — has so far refused.

On Friday, Zuckerberg said in a post that he would review “potential options for handling violating or partially-violating content aside from the binary leave-it-up or take-it-down decisions”

“I know many of you think we should have labeled the President’s posts in some way last week,” he wrote. “Our current policy is that if content is actually inciting violence, then the right mitigation is to take that content down — not let people continue seeing it behind a flag. There is no exception to this policy for politicians or newsworthiness.”
"""

tfidf = TfidfVectorizer(stop_words='english')
data = tfidf.fit_transform([doc1,doc2,doc3])

vector1 = tfidf.transform([doc1]).toarray()
vector2 = tfidf.transform([doc2]).toarray()
vector3 = tfidf.transform([doc3]).toarray()

cosine1 = cosine_similarity(vector1, vector2)
print("Cosine similarity between doc1 and doc2:", cosine1)
cosine2 = cosine_similarity(vector2, vector3)
print("Cosine similarity between doc3 and doc3:", cosine2)
cosine3 = cosine_similarity(vector1, vector3)
print("Cosine similarity between doc1 and doc3:", cosine3)

Cosine similarity between doc1 and doc2: [[0.23784716]]
Cosine similarity between doc3 and doc3: [[0.06526311]]
Cosine similarity between doc1 and doc3: [[0.06431176]]


# After applying text preprocessing

In [6]:
import nltk
import re
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer
from string import punctuation
from autocorrect import spell

snowball_stemmer = SnowballStemmer('english')
wordnet_lemmatizer = WordNetLemmatizer()

class Preprocess:
    def __int__(self):
        pass

    def autospell(self,text):
        """
        correct the spelling of the word.
        """
        spells = [spell(w) for w in (nltk.word_tokenize(text))]
        return " ".join(spells)

    def to_lower(self,text):
        """
        :param text:
        :return:
            Converted text to lower case as in, converting "Hello" to "hello" or "HELLO" to "hello".
        """
        return text.lower()

    def remove_numbers(self,text):
        """
        take string input and return a clean text without numbers.
        Use regex to discard the numbers.
        """
        output = ''.join(c for c in text if not c.isdigit())
        return output

    def remove_punct(self,text):
        """
        take string input and clean string without punctuations.
        use regex to remove the punctuations.
        """
        return ''.join(c for c in text if c not in punctuation)

    def remove_Tags(self,text):
        """
        take string input and clean string without tags.
        use regex to remove the html tags.
        """
        cleaned_text = re.sub('<[^<]+?>', '', text)
        return cleaned_text

    def sentence_tokenize(self,text):
        """
        take string input and return list of sentences.
        use nltk.sent_tokenize() to split the sentences.
        """
        sent_list = []
        for w in nltk.sent_tokenize(text):
            sent_list.append(w)
        return sent_list

    def word_tokenize(self,text):
        """
        :param text:
        :return: list of words
        """
        return [w for sent in nltk.sent_tokenize(text) for w in nltk.word_tokenize(sent)]

    def remove_stopwords(self,sentence):
        """
        removes all the stop words like "is,the,a, etc."
        """
        stop_words = stopwords.words('english')
        return ' '.join([w for w in nltk.word_tokenize(sentence) if not w in stop_words])

    def stem(self,text):
        """
        :param word_tokens:
        :return: list of words
        """
        stemmed_word = [snowball_stemmer.stem(word) for sent in nltk.sent_tokenize(text)for word in nltk.word_tokenize(sent)]
        return " ".join(stemmed_word)

    def lemmatize(self,text):
        lemmatized_word = [wordnet_lemmatizer.lemmatize(word)for sent in nltk.sent_tokenize(text)for word in nltk.word_tokenize(sent)]
        return " ".join(lemmatized_word)


    def preprocess(self,text):
        lower_text = self.to_lower(text)
        sentence_tokens = self.sentence_tokenize(lower_text)
        word_list = []
        for each_sent in sentence_tokens:
            lemmatizzed_sent = self.lemmatize(each_sent)
            clean_text = self.remove_numbers(lemmatizzed_sent)
            clean_text = self.remove_punct(clean_text)
            clean_text = self.remove_Tags(clean_text)
            clean_text = self.remove_stopwords(clean_text)
            word_tokens = self.word_tokenize(clean_text)
            for i in word_tokens:
                word_list.append(i)
        return word_list

In [7]:
pr = Preprocess()
doc1_cleaned = " ".join(pr.preprocess(doc1))
doc2_cleaned = " ".join(pr.preprocess(doc2))
doc3_cleaned = " ".join(pr.preprocess(doc3))

tfidf = TfidfVectorizer(stop_words='english')
tfidf.fit_transform([doc1_cleaned, doc2_cleaned, doc3_cleaned])

vector1 = tfidf.transform([doc1_cleaned]).toarray()
vector2 = tfidf.transform([doc2_cleaned]).toarray()
vector3 = tfidf.transform([doc3_cleaned]).toarray()

cosine1 = cosine_similarity(vector1, vector2)
print("Cosine similarity between doc1 and doc2:", cosine1)
cosine2 = cosine_similarity(vector2, vector3)
print("Cosine similarity between doc3 and doc3:", cosine2)
cosine3 = cosine_similarity(vector1, vector3)
print("Cosine similarity between doc1 and doc3:", cosine3)

Cosine similarity between doc1 and doc2: [[0.28282452]]
Cosine similarity between doc3 and doc3: [[0.07791204]]
Cosine similarity between doc1 and doc3: [[0.07407567]]


<b>NOTE:</b>
As you can see there's slight score change with and without the pre-processing. <br>
Before pre-processing:<br>
Cosine similarity between doc1 and doc2: [[0.23784716]]<br>
Cosine similarity between doc3 and doc3: [[0.06526311]]<br>
Cosine similarity between doc1 and doc3: [[0.06431176]]<br><br>
After pre-processing: <br>
Cosine similarity between doc1 and doc2: [[0.28282452]]<br>
Cosine similarity between doc3 and doc3: [[0.07791204]]<br>
Cosine similarity between doc1 and doc3: [[0.07407567]]<br>